pkg/store/filestore.go

```diff
@@ -0,0 +1,415 @@
+		return nil, err
+	}
+
+	path := filepath.Join(append([]string{vs.dir}, key...)...)
+
+	st, err := os.Stat(path)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return nil, errors.Join(ErrNotFound, fmt.Errorf("%q does not exist", filepath.Join(key...)))
+		}
+
+		return nil, errors.Join(ErrSystemFailure, err)
+	}
+
+	if st.IsDir() {
+		return nil, errors.Join(ErrFaultyImplementation, fmt.Errorf("%q is a directory and cannot be read as a file", path))
+	}
+
+	content, err := os.ReadFile(filepath.Join(append([]string{vs.dir}, key...)...))
+	if err != nil {
+		return nil, errors.Join(ErrSystemFailure, err)
+	}
+
+	return content, nil
+}
+
+func (vs *fileStore) Exists(key ...string) (bool, error) {
+	if err := validateAllPathComponents(key...); err != nil {
+		return false, err
+	}
+
+	path := filepath.Join(append([]string{vs.dir}, key...)...)
+
+	_, err := os.Stat(filepath.Join(path))
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return false, nil
+		}
+
+		return false, errors.Join(ErrSystemFailure, err)
+	}
+
+	return true, nil
+}
+
+func (vs *fileStore) Set(data []byte, key ...string) error {
+	if vs.locked == nil {
+		return errors.Join(ErrFaultyImplementation, fmt.Errorf("operations on the store must use locking"))
+	}
+
+	if err := validateAllPathComponents(key...); err != nil {
+		return err
+	}
+
+	fileName := key[len(key)-1]
+	parent := vs.dir
+
+	if len(key) > 1 {
+		parent = filepath.Join(append([]string{parent}, key[0:len(key)-1]...)...)
+		err := os.MkdirAll(parent, vs.dirPerm)
+		if err != nil {
+			return errors.Join(ErrSystemFailure, err)
+		}
+	}
+
+	dest := filepath.Join(parent, fileName)
+	st, err := os.Stat(dest)
+	if err == nil {
+		if st.IsDir() {
+			return errors.Join(ErrFaultyImplementation, fmt.Errorf("%q is a directory and cannot be written to", dest))
+		}
+	}
+
+	return atomicWrite(parent, fileName, vs.filePerm, data)
+}
+
+func (vs *fileStore) List(key ...string) ([]string, error) {
+	if vs.locked == nil {
+		return nil, errors.Join(ErrFaultyImplementation, fmt.Errorf("operations on the store must use locking"))
+	}
+
+	// Unlike Get, Set and Delete, List can have zero length key
+	for _, k := range key {
+		if err := validatePathComponent(k); err != nil {
+			return nil, err
+		}
+	}
+
+	path := filepath.Join(append([]string{vs.dir}, key...)...)
+
+	st, err := os.Stat(path)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return nil, errors.Join(ErrNotFound, err)
+		}
+
+		return nil, errors.Join(ErrSystemFailure, err)
+	}
+
+	if !st.IsDir() {
+		return nil, errors.Join(ErrFaultyImplementation, fmt.Errorf("%q is not a directory and cannot be enumerated", path))
+	}
+
+	dirEntries, err := os.ReadDir(path)
+	if err != nil {
+		return nil, errors.Join(ErrSystemFailure, err)
+	}
+
+	entries := []string{}
+	for _, dirEntry := range dirEntries {
+		entries = append(entries, dirEntry.Name())
+	}
+
+	return entries, nil
+}
+
+/*
```

Comment:
Will remove before merging.
This is not used.

Reply:
This should be removed now?